### ISSUE TITLE
Consistent use of (SUB)PRIORITY and NUMOBS_MORE throughout fiber assignment

### ISSUE THREAD

**Issue by kdawson1000** [desihub/fiberassign#140]
SUBPRIORITY is currently used as a tie-breaker when science targets of equal priority compete for the same fiber.  However, it is not being used as a tie-breaker when std star or sky targets get assigned.  

In the case that the number of available fibers is less than the number of required std star/sky targets, the code randomly displaces the lowest priority science targets with calibration targets.  Adding the SUBPRIORITY as a tie-breaker will allow better control of the sample assigned fibers.  This will be particularly important for SV when we have a set goal of NOBS=4 for each science target.

**Comment by kdawson1000** [comment]
Stephen's point is very reasonable.  Is SUBPRIORITY really limited to (float, 0-1)?  If so, there will be no concern about pushing a target with PRIORITY+SUBPRIORITY so that it bumps a target from a higher priority class.

**Comment by moustakas** [comment]
Yes, `SUBPRIORITY` is limited to [0.0, 1.0):
https://github.com/desihub/desitarget/blob/master/py/desitarget/io.py#L205

**Comment by tskisner** [comment]
I was just looking at this code today:

https://github.com/desihub/fiberassign/blob/master/src/global.cpp#L217

And thinking that those compound conditionals should be replaced with a simple calculation of a single number that could be simply used to rank the available targets for each tile filber.  Something like:

pr = (double)priority * 1.0e10 + (double)obs_remaining * 1.0e8 + subpriority;

Where those multiplicative factors were just chosen to "use more of the bits" available in the double precision number and still allow a large range for all three quantities.  If subpriority is only 0-1, then these numbers don't need to be so big.

Note that I was still going to drag around both priority and subpriority- that makes it easier to do things like above, where we want the 3 "sub-parts" of the overall priority ranked in a particular order (priority, obs remaining, subpriority).

**Comment by sbailey** [comment]
Interesting.  Those lines of code highlight that needing more observations trumps subpriority (as would your proposed replacement).  Though I'm not sure either one is particularly clear without the comment above it to explain the reasoning.

I'm also not sure which assignment is best: if LyA QSO A has 3 observation and B has 2 observations and we only have 1 fiber left and it covers both, would we rather get a 4th observations of A or a 3rd observation of B?

**Comment by tskisner** [comment]
In addition to splitting out the I/O, my focus in the refactor for #136 is just cleaning up the code and making it easier to determine what is going on without changing the underlying algorithm.  But yes, good to consider what the desired behavior should be.

**Comment by kdawson1000** [comment]
We have to ask the TS group and the Lyman-alpha group what they want in the case of two high priority objects competing for the same fiber when they both have NOBS_MORE>1.  The obvious cases are:

(a) two quasar targets, one is confirmed z>2.1 with only one epoch and the other has not been observed
(b) two quasar targets, both are confirmed z>2.1 and both have NOBS_MORE>1
(c) two LRG targets, one is z>20 and has one observation, the other is z>20 and has no observations
(d) two BGS targets, one is observed but not confirmed, the other is not observed (I still think we should impose mag limits for NOBS)

**Comment by kdawson1000** [comment]
yes, definitely.  I am happy to gather this information from the science teams and get it recorded on the TS wiki.

**Comment by kdawson1000** [comment]
We have feedback from the galaxy groups.  We should use the following scheme for fiber collisions:

(c) two LRG targets, one is z>20 and has one observation, the other is z>20 and has no observations.  Do we continue to observe the first target or do we prioritize the new target?  ***ACTION:  Continue to observe the first one to 2X depth.  LOGIC:  Completeness of z>20 LRGs is supposed to be achieved by 2-pass observations.  Dividing time between the two reduces the likelihood that any one object gets the two passes.

(d)  two LRG targets, one is z>20 and has one observation, the other is z<20 and has no observations.  Do we continue to observe the first target or do we prioritize the new target?  ***ACTION:  Continue to observe the first one to 2X depth.  LOGIC:  Completeness of z>20 LRGs is supposed to be achieved by 2-pass observations.  Two-pass observations likely to be high redshift where n(z) is dropping for LRG.  Invest time in object that likely constitutes a larger fraction of the number density at its redshift.

(e) two BGS targets of equal priority, one is observed but not confirmed, the other is not observed (I still think we should impose mag limits for NOBS).  ***ACRION:  observe the new target.  LOGIC:  I think this is the BGS proposal anyway

undecided, what happens in case (e) if the two BGS targets have different priority?

**Comment by kdawson1000** [comment]
We have feedback from the lyman-alpha forest group, the first scenario is decided:
    (a) two quasar targets, one is confirmed z>2.1 with only one epoch and the other has not been observed.  Do we choose deeper observations of the known z>2.1 quasar or do we choose to increase sample size by taking an epoch of the unknown quasar target?  ***ACTION:  assign fiber to the known quasar for one additional epoch (obtaining two observations) and then assign fibers to the unknown object for third pass.    LOGIC: close quasar pairs are quite interesting, and that the difference between having 3 or 4 passes for a Lyman alpha quasar is not that big.  This way, if the new target turns out to also be a Lyman alpha quasar, at least you'll get two passes on each.

**Comment by kdawson1000** [comment]
The second scenario for lyman-alpha needs more thought.  Specifically, how nuanced do we want our criteria to be for deciding follow-up observations?  One could use a script provided by the Lyman-alpha WG as part of the MTL process and leave the complexity to those folks.  The thought process is below:

 (b) two quasar targets, both are confirmed z>2.1 and both have NOBS_MORE>1.  Do we choose to observe one quasar for all remaining epochs or do we divide the remaining fiber assignments between the two quasars?  ****ACTION:  aim for equal number of fiber assignments with remaining layers, but depending on value of quasar target.  Let Lyman-alpha group establish algorithm.  LOGIC:  deeper spectra hit diminishing returns after two observations, but every quasar has a different value depending on its likelihood of being a quasar, S/N, and redshift.  Better to divide the available fibers between the two targets rather than push one target into diminishing returns.

**Comment by kdawson1000** [comment]
I missed another Lyman-alpha situation.  Are we using information from BOSS/eBOSS in the target selection and prioritization of quasar targets?  The case to consider is:

(f) on the first pass, we have a BOSS z>2.1 quasar, and a quasar target that was not observed in BOSS (this might also be an issue with LRGs if we decided to reobserve them). PROPOSAL: we observe the non-BOSS quasar in the first pass. LOGIC: If it was a BOSS quasar it is probably bright enough that we will get a good spectrum with only two passes. This of course assumes that in (b) we decide to split 2 and 2.

**Comment by geordie666** [comment]
Hi Kyle,

Thanks...this is useful. Can someone with fiberassign expertise indicate
whether we store the knowledge that something is a collided pair in the
context of reobserving it on additional passes?

In targeting (MTL; merged target list), we can change the priority of *any
individual target* based on the current spectroscopic knowledge for that
individual target. But, I don't think that MTL can currently change the
priority of targets conditioned on the fact that the target is in a
collided pair. Perhaps that is handled in fiberassign?

Cheers,

Ad.

**Comment by rncahn** [comment]
We don't store information on past potential collisions because they are
specific to a particular target and most targets are observed only once and
because the collisions are also specific to the location of the center of
the field.

**Comment by tskisner** [comment]
Revisiting this issue in light of recent code changes:

1.  SUBPRIORITY is essentially a random number on the interval [0,1), but a random number generated at the catalog creation time, which makes it independent of fiberassign (which is good).  Sky and standards have a subpriority and this is used consistently throughout the code.

2.  PRIORITY is an integer.  There are no restrictions on the number of unique priority values, **but** the set of these values across all targets is used when bumping science targets for standards or sky.  This list of all available PRIORITY values is looped over in order from low to high and for each value, science targets of that "priority class" are considered for bumping.  Sky and "normal" standards have priority == 0.  Standards which are also science targets will have priority > 0.

3.  The algorithm to determine which target is "best" is currently dependent on whether the target is a science target vs a standard or sky.  This "decision" is isolated to one place in the code:

    https://github.com/desihub/fiberassign/blob/master/src/assign.cpp#L1426

As noted above, for science targets, this algorithm includes the obs remaining, not just the priority + subpriority.

I'm not sure that anyone can be fully confident on the impact of a given choice without running a test for each modification and comparing the resulting assignment efficiencies.

**Comment by kdawson1000** [comment]
I have a somewhat related question:  when you run the replacement of science targets with sky targets to meet the required density of calibration fibers, do you loop on the subpriority of the science targets or on the subpriority of the sky targets.  In an earlier email exchange, I inferred that you were basing the order of replacement on the subpriority of the sky targets.  It would be better to base the order of replacement on the priority+subpriority of the science targets.

**Comment by sbailey** [comment]
Agreed with @kdawson1000 .  Bumping science targets for sky and std should be done in order of science target priority+subpriority, not in order of sky/std subpriority, nor in order of fiber number.

@tskisner: this ticket is spinning into multiple related topics.  We should resolve it soon, or otherwise split them out into separate tickets for clarity:
* Use of SUBPRIORITY instead of fiber number to order original STD and SKY assignments (original post; I'm not sure if this was every confirmed or if it was an assertion/question).  Possibly related to #140 sky non-uniformity issue.
* Use of science target PRIORITY+SUBPRIORITY when bumping for STD and SKY
* When should NUMOBS_MORE trump SUBPRIORITY (yes for QSO LyA and SV, no for others)

**Comment by tskisner** [comment]
There seems to be persistent confusion.  Hopefully my in-progress work on the documentation will help.  

1.  The initial assignment of unused fibers has always looped over fibers in order based on the total priority of the targets available to the fibers.  

2.  When **redistributing** science targets to future tile/fibers, both the legacy and current code loop over fibers in ID order.  This needs to be changed (see #179).

3.  When bumping science targets for standards or sky, both legacy and current code use the priority of the standard or sky to set the order.  The fibers are then looped over for each priority class from low to high.  This means that the fiber looping order will be random within a tile and this means that targets in a given priority class will be bumped randomly based on location on the tile.  Like this pseudocode:
  ```
  foreach priority class (e.g. 1500 ... 3200):
      foreach **sky or standard** in subpriority (i.e. random) order:
          foreach fiber that can reach this sky or standard:
              if fiber is assigned to target of specified priority class:
                  bump science target
                  assign sky / standard
                  try to reassign bumped science target
  ```
  Or if you prefer the actual code:
  https://github.com/desihub/fiberassign/blob/master/src/assign.cpp#L843

4.  The details of how to determine which target from a list is "best" is encapsulated in a single function where the comparison is implemented based on type of target, priority, subpriority, and obs remaining.  That one place is where you can change the behavior.

I'll use this discussion as a guide to improving the docs.

**Comment by sbailey** [comment]
@tskisner thanks for succinctly describing the current algorithms.  Unfortunately all of them are somewhat different from what we ideally want.  I'm spinning these into separate tickets so that we aren't tracking multiple separate concepts and fixes in a single ticket.

> 1. The initial assignment of unused fibers has always looped over fibers in order based on the total priority of the targets available to the fibers.

See #183 

> 2. When **redistributing** science targets to future tile/fibers, both the legacy and current code loop over fibers in ID order.  This needs to be changed (see #179).

Agreed; let's continue discussion in #179.

> 3. When bumping science targets for standards or sky, both legacy and current code use the priority of the standard or sky to set the order.  The fibers are then looped over for each priority class from low to high.  This means that the fiber looping order will be random within a tile and this means that targets in a given priority class will be bumped randomly based on location on the tile.  

See #184 

> 4. The details of how to determine which target from a list is "best" is encapsulated in a single function where the comparison is implemented based on type of target, priority, subpriority, and obs remaining.  That one place is where you can change the 
behavior.

This is a different from the original title of this ticket but there are multiple long comments in the history about special cases for NUMOBS_MORE that I haven't (yet?) moved into a new ticket, but there is substantive work to be done to update the logic to handle these cases.

**Comment by sbailey** [comment]
I'm re-posting some comments from Kyle from months ago which may have gotten lost in the multiple-topics discussed in this ticket.  This is about when should SUBPRIORITY trump NUMOBS_MORE or vice-a-versa.  The current code that decides this is https://github.com/desihub/fiberassign/blob/master/src/assign.cpp#L1411, where NUMOBS_MORE trumps SUBPRIORITY for all science targets.  This is generally what we want for LyA QSO, but the opposite of what we want for LRGs.  How should that distinction be handled in the code?

# GQC

<hr/>

**EXAMPLE**: Two LRG targets, one is z>20 and has one observation, the other is z>20 and has no observations. Do we continue to observe the first target or do we prioritize the new target?

**ACTION**: Continue to observe the first one to 2X depth.

**REASON**: Completeness of z>20 LRGs is supposed to be achieved by 2-pass observations. Dividing time between the two reduces the likelihood that any one object gets the two passes.

<hr/>

**EXAMPLE**: two LRG targets, one is z>20 and has one observation, the other is z<20 and has no observations. Do we continue to observe the first target or do we prioritize the new target?

**ACTION**: Continue to observe the first one to 2X depth.

**REASON**: Completeness of z>20 LRGs is supposed to be achieved by 2-pass observations. Two-pass observations likely to be high redshift where n(z) is dropping for LRG. Invest time in object that likely constitutes a larger fraction of the number density at its redshift.

<hr/>

**EXAMPLE** two BGS targets of equal priority, one is observed but not confirmed, the other is not observed (I (Kyle) still think we should impose mag limits for NOBS). 

**ACTION**: observe the new target.

**REASON**: I (Kyle) think this is the BGS proposal anyway

# Lyman-alpha

**Kyle editorial**: These scenarios for lyman-alpha needs more thought. Specifically, how nuanced do we want our criteria to be for deciding follow-up observations? One could use a script provided by the Lyman-alpha WG as part of the MTL process and leave the complexity to those folks. The thought process is below.

**Stephen Editorial**:  it can't necessarily be handled only at the MTL level, since fiberassign itself may have 4 opportunities to place 2 targets that are each requesting 4 observations, and it needs to make decisions before taking more observations and updating the MTL.  Having complex decision logic deep inside fiberassign itself is yucky.

<hr/>

**EXAMPLE**: two quasar targets, one is confirmed z>2.1 with only one epoch and the other has not been observed. Do we choose deeper observations of the known z>2.1 quasar or do we choose to increase sample size by taking an epoch of the unknown quasar target?

**(Requested) ACTION**: assign fiber to the known quasar for one additional epoch (obtaining two observations) and then assign fibers to the unknown object for third pass.

**REASON**: close quasar pairs are quite interesting, and that the difference between having 3 or 4 passes for a Lyman alpha quasar is not that big. This way, if the new target turns out to also be a Lyman alpha quasar, at least you'll get two passes on each.

<hr/>

**EXAMPLE**: two quasar targets, both are confirmed z>2.1 and both have NOBS_MORE>1. Do we choose to observe one quasar for all remaining epochs or do we divide the remaining fiber assignments between the two quasars?

**(Requested) ACTION**: aim for equal number of fiber assignments with remaining layers, but depending on value of quasar target. Let Lyman-alpha group establish algorithm.

**REASON**: deeper spectra hit diminishing returns after two observations, but every quasar has a different value depending on its likelihood of being a quasar, S/N, and redshift. Better to divide the available fibers between the two targets rather than push one target into diminishing returns.

<hr/>

**EXAMPLE**: on the first pass, we have a BOSS z>2.1 quasar, and a quasar target that was not observed in BOSS (this might also be an issue with LRGs if we decided to reobserve them).

**Kyle PROPOSAL**: we observe the non-BOSS quasar in the first pass.

**REASON**: If it was a BOSS quasar it is probably bright enough that we will get a good spectrum with only two passes. This of course assumes that in (b) we decide to split 2 and 2.

**Comment by andluizsouza** [comment]
Hi people.

I would like to know what is exactly the current status of this ticket. 

I am studying those fiber assignment tutorials. This [one](https://github.com/desihub/tutorials/blob/master/FiberAssign.ipynb) reports the current algorithm prioritizes observing more targets before getting more observations of a previously observed target. 

There is a TODO task: **_to add an option to fiberassign to give the flexibility to finish targets before starting new ones_.**

Was this task already finished? If not so, I am interested in helping.
Thanks a lot!

**Comment by forero** [comment]
Hi @andluizsouza. The TODO item you mention has not been addressed.

**Comment by andluizsouza** [comment]
Hi people!

I would like to return to this discussion.

Does anyone have a tip about how I could set `SUBPRIORITY` flags between the epochs (the calls to fiber assignments) knowing if that target was or not assigned previously?
Can I modify the updated _mtl.fits_ that is passed back to fiber assignment for the next epoch?

I have read this tutorial https://fiberassign.readthedocs.io/en/latest/usage.html# but I didn't find anything about this.

Thanks a lot!

**Comment by tskisner** [comment]
The PRIORITY is an integer which describes the relative priority of different target types.  SUBPRIORITY is an intentionally random float64.  The total priority (PRIORITY + SUBPRIORITY) is used when setting the order of considering targets for assignment.  When assigning science targets, obs remaining is considered as per the code link in this thread.

We should not be making SUBPRIORITY non-random.  Instead, we should have separate bits of DESI_TARGET or some other column that indicates whether the target should be assigned depth first, breadth first, or ignore the obs remaining and just use total priority.  This way the per-target behavior is controlled outside of fiberassign.

**Comment by andluizsouza** [comment]
Thanks, @tskisner!
I'd like to rank Lya QSOs with different priorities to the fiber assignment according to an external redshift classifier. So could I split my original mtl file into some target files, each one with its priority number and number of observations, and compile all of them together?

**Comment by geordie666** [comment]
It sounds like you want to modify how the MTL assigns priorities based on information on the spectroscopic redshift. That logic occurs here:

https://github.com/desihub/desitarget/blob/aad6fd71797f349bcb8db14e8b4d4ed044b7a6a8/py/desitarget/targets.py#L481-L495

Your "ranking" would actually be an update to the "MORE_ZGOOD" value of priority. As this is the highest priority in the targeting yaml file:

https://github.com/desihub/desitarget/blob/aad6fd71797f349bcb8db14e8b4d4ed044b7a6a8/py/desitarget/data/targetmask.yaml#L164-L194

you could just add integer values to the "MORE_ZGOOD" priority value.

However: Continuous ranking is non-standard behavior for how we've set up `desitarget`. It requires information about priorities outside of the yaml file. We looked at such rankings in the past for BOSS and found the gains were negligible for LyA quasars. So, I think there'd be a reasonably high bar to including a continuous ranking.

That doesn't mean that we can't plug in an external classifier with an "either/or" rank. As in "my preferred redshift classifier says this is a LyA quasar, we should give it the `MORE_ZGOOD` priority". Or, have, say, 3-4 priorities (`MORE_ZGOOD`, `LOVAL_ZGOOD`, `MIDVAL_ZGOOD`, `HIVAL_ZGOOD`).

**Comment by tskisner** [comment]
Thanks @geordie666 for making my in-progress comment redundant :-)

Obviously for testing, one could read in a MTL file and manually set PRIORITY values for targets to a higher or lower value.  However, this is not a behavior we want to use in production.  Thanks for spelling out the details.

**Comment by tskisner** [comment]
Also- I think it would be great to move the depth-first / bread-first / neither decision out of fiberassign and into some input target property.  This is a tri-state, so would need two bits.  I know DESI_TARGET is getting a "bit crowded" (haha!) but we could use two bits there or put it in some other single-byte column of the input target files (e.g "STRATEGY" or "ASGNALGO" or some other clever 8 character name).

Obviously needs more discussion, but this would be a trivial change to fiberassign once we have consensus on the external details.

**Comment by andluizsouza** [comment]
Thanks a lot, @geordie666 and @tskisner.
I'm going to follow your directions and to think about how to apply them out of the fiber assignment algorithm firstly.

**Comment by andluizsouza** [comment]
More one question: what does `ZWARN` in zcat (input into the calc_priority function) mean exactly?

**Comment by geordie666** [comment]
In general, if `ZWARN` is zero, then the spectral classifier produced a good, believable fit. If `ZWARN` is a non-zero bit, then there is a potential issue with the fit to the spectrum.

For example, for the `redrock` classifier, here are the possible values of `ZWARN`:

https://github.com/desihub/redrock/blob/335d26eb8da0b56f22f9340795a0cf35f3ab0b1f/py/redrock/zwarning.py#L14-L25

**Comment by tskisner** [comment]
This is a massive issue that bounced around in scope.  As of the code version in #258, we have:

- Targets have a "total priority" that is a combination of PRIORITY, SUBPRIORITY, and NUMOBS_MORE.  Currently this total priority is calculated the same for all targets (using a "breadth first" strategy).

- Within a single target class (with the same PRIORITY), it would be very easy to use a "depth first" strategy instead.  This switch could be controlled on a per-target basis using one bit of DESI_TARGET or similar.

I think we should close this issue and open a new one with that feature suggestion as a question for discussion.  Sound ok @sbailey?

**Comment by tskisner** [comment]
As previously stated, this original point of this issue is no longer relevant.  The remaining question of how to prioritize targets **within** a single class (with the same PRIORITY) based on the observations remaining is covered by #196 .  Closing this now
